[grid][java]: session-timeout set connection timeout in RemoteNode

VietND96 opened the pull request:
## **User description**
**Thanks for contributing to Selenium!**
**A PR well described will help maintainers to quickly review and merge it**

Before submitting your PR, please check our [contributing](https://github.com/SeleniumHQ/selenium/blob/trunk/CONTRIBUTING.md) guidelines.
Avoid large PRs, help reviewers by making them as simple and short as possible.


<!--- Provide a general summary of your changes in the Title above -->

### Description
<!--- Describe your changes in detail -->

### Motivation and Context
Try something for https://github.com/SeleniumHQ/selenium/issues/13340
`session-timeout` is an attribute of all node types and can get in part of `NodeStatus`
Use `session-timeout` to set connection timeout and read timeout for HttpClient config in `RemoteNode`

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have read the [contributing](https://github.com/SeleniumHQ/selenium/blob/trunk/CONTRIBUTING.md) document.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
<!--- Provide a general summary of your changes in the Title above -->


___

## **Type**
enhancement


___

## **Description**
- Added `sessionTimeout` to `NodeStatus` and updated JSON handling.
- Enhanced various node types and distributor classes to include and manage `sessionTimeout`.
- Updated tests across multiple classes to incorporate `sessionTimeout` in node configurations and assertions.



___



## **Changes walkthrough**
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement
</strong></td><td><details><summary>9 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>NodeStatus.java</strong><dd><code>Add session timeout to NodeStatus</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
java/src/org/openqa/selenium/grid/data/NodeStatus.java

<li>Added <code>sessionTimeout</code> field to <code>NodeStatus</code> class.<br> <li> Updated constructors and JSON serialization/deserialization to handle <br><code>sessionTimeout</code>.<br>


</details>
    

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/13854/files#diff-04bf71dc8f94e5e1e3620bb9b03ea6162519e7adefa5dbec7a420a5c88dfc1d8">+22/-1</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>AddNode.java</strong><dd><code>Include session timeout in node creation</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
java/src/org/openqa/selenium/grid/distributor/AddNode.java

<li>Modified <code>execute</code> method to include <code>sessionTimeout</code> when creating a new <br>node.<br>


</details>
    

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/13854/files#diff-a6491ce1d5f18cd0efbe3d13311b583a9690fde1d98c27233655b2526097a018">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>GridModel.java</strong><dd><code>Handle session timeout in GridModel</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
java/src/org/openqa/selenium/grid/distributor/GridModel.java

- Updated `rewrite` and `amend` methods to handle `sessionTimeout`.



</details>
    

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/13854/files#diff-72eb65833e6f3833e365d678e81969b2ecc3210df5c9b1ef0e044086ae1e6b53">+2/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>LocalDistributor.java</strong><dd><code>Adjust node registration to include session timeout</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
java/src/org/openqa/selenium/grid/distributor/local/LocalDistributor.java

<li>Adjusted <code>register</code> method to include <code>sessionTimeout</code> in node creation.<br>


</details>
    

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/13854/files#diff-024d7dccb0a4ff87c444dc3549032e7c6b2595f582f4f267cdf48c7f8c2f87c1">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>Node.java</strong><dd><code>Add session timeout management in Node base class</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
java/src/org/openqa/selenium/grid/node/Node.java

<li>Added <code>sessionTimeout</code> to the Node class and updated constructors.<br> <li> Provided a getter for <code>sessionTimeout</code>.<br>


</details>
    

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/13854/files#diff-9724970402f0d888442caa09bf30d6f8074edde726dae3e98a52eda4f2ed8462">+12/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>OneShotNode.java</strong><dd><code>Manage session timeout in OneShotNode</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
java/src/org/openqa/selenium/grid/node/k8s/OneShotNode.java

<li>Included <code>sessionTimeout</code> in the constructor and during node creation.<br>


</details>
    

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/13854/files#diff-e2faca1f36057dbd7099e2881d61cd68301db4b5f6c2b1ac616b2189979f388e">+4/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>LocalNode.java</strong><dd><code>Update LocalNode to handle session timeout</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
java/src/org/openqa/selenium/grid/node/local/LocalNode.java

- Updated constructors to handle `sessionTimeout`.



</details>
    

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/13854/files#diff-edc2e1943a5d89fa45151a9d214cc608551c61101f237ca9a5993d6dd157b57b">+7/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>RemoteNode.java</strong><dd><code>Enhance RemoteNode with session timeout configuration</code>&nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
java/src/org/openqa/selenium/grid/node/remote/RemoteNode.java

<li>Updated constructors to handle <code>sessionTimeout</code>.<br> <li> Configured HTTP client to use <code>sessionTimeout</code> for read timeout.<br>


</details>
    

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/13854/files#diff-7d02354cb02076c23a0bda07882c1fd634c19eee19be14b96a4b77cd796cde75">+8/-3</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>GridRedisClient.java</strong><dd><code>Include session timeout in Redis client node retrieval</code>&nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
java/src/org/openqa/selenium/redis/GridRedisClient.java

<li>Updated <code>getNode</code> method to include <code>sessionTimeout</code> in the node status.<br>


</details>
    

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/13854/files#diff-88925238f960e0340072ab4102110848382c8b519d019ceae839f8a2eb965a4b">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></details></td></tr><tr><td><strong>Tests
</strong></td><td><details><summary>4 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>NodeStatusTest.java</strong><dd><code>Update NodeStatus tests to include session timeout</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
java/test/org/openqa/selenium/grid/data/NodeStatusTest.java

- Updated test to include `sessionTimeout` in `NodeStatus`.



</details>
    

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/13854/files#diff-593943d5e5fd01f7d03d2e32f8aea53c9d25d46996296e231b2afb1eb19925ed">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>AddingNodesTest.java</strong><dd><code>Update distributor tests for session timeout handling</code>&nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
java/test/org/openqa/selenium/grid/distributor/AddingNodesTest.java

<li>Updated tests to handle <code>sessionTimeout</code> in node status events and <br>status retrieval.<br>


</details>
    

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/13854/files#diff-24b972fe88b1e515a226b8c3b1cd68eb9a0801a5584d83321c820425fb2cd254">+2/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>DefaultSlotSelectorTest.java</strong><dd><code>Update slot selector tests to include session timeout</code>&nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
java/test/org/openqa/selenium/grid/distributor/selector/DefaultSlotSelectorTest.java

- Adjusted node creation in tests to include `sessionTimeout`.



</details>
    

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/13854/files#diff-c3acffd6266dc62129064038e49aea1f6dfaf453dbd0aac738d281697babc892">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>NodeTest.java</strong><dd><code>Update Node tests to handle session timeout</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
java/test/org/openqa/selenium/grid/node/NodeTest.java

- Updated node creation in tests to handle `sessionTimeout`.



</details>
    

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/13854/files#diff-7fc2523b1e86e098543cf0def812b38114ce596668c61b8259ead0e1dba56a10">+4/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></details></td></tr></tr></tbody></table>

___

> ✨ **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

VietND96 commented:
@diemol, can you also review this one?

diemol commented:
@VietND96 can you please update your fork?

diemol commented on java/src/org/openqa/selenium/grid/node/Node.java:
```diff
@@ -116,13 +117,20 @@ public abstract class Node implements HasReadyState, Routable {
   protected boolean draining;
 
   protected Node(Tracer tracer, NodeId id, URI uri, Secret registrationSecret) {
+    this(tracer, id, uri, registrationSecret, Duration.ofSeconds(300));
```
Why do we need the default value for `sessionTimeout` here? Can't we change the signature? Does it break something?

VietND96 replied:
This constructor is used by 1 unit test, so I updated a specific value in the test alone and removed this constructor.

VietND96 commented on java/src/org/openqa/selenium/grid/node/remote/RemoteNode.java:
```diff
@@ -83,13 +86,15 @@ public RemoteNode(
-    this.client =
-        Require.nonNull("HTTP client factory", clientFactory).createClient(fromUri(externalUri));
+    ClientConfig clientConfig =
+        defaultConfig().readTimeout(this.getSessionTimeout()).baseUrl(fromUri(externalUri));
```
@diemol, do you think connectionTimeout() also need here?

diemol replied:
I don't think so. Connection timeout is used for the initial connection, and in theory all components are in easily reachable network(s).

VietND96 replied:
Sure, so we set `readTimeout` only.

diemol reviewed (approved):
Need to update the fork.

VietND96 commented:
@diemol, I updated the fork. Also, I found that the test is used to define a CustomNode, and it can be added. So, I updated the CustomNode with attr `sessionTimeout,` which can be set when creating. In the test, it asserts that `getSessionTimeout()` is working. Check if the update for the test makes sense.
